File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Proxy.VirtualProxy/FormVirtualProxy.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Threading
Imports System.Net

Namespace DoFactory.HeadFirst.Proxy.VirtualProxy
	''' <summary>
	''' Summary description for FormVirtualProxy.
	''' </summary>
	Public Class FormVirtualProxy
		Inherits System.Windows.Forms.Form
		Private WithEvents buttonTestImageProxy As System.Windows.Forms.Button
		Private WithEvents pictureBox As System.Windows.Forms.PictureBox
		Private WithEvents label1 As System.Windows.Forms.Label
		''' <summary>
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.Container = Nothing

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.buttonTestImageProxy = New System.Windows.Forms.Button()
			Me.pictureBox = New System.Windows.Forms.PictureBox()
			Me.label1 = New System.Windows.Forms.Label()
			CType(Me.pictureBox, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' buttonTestImageProxy
			' 
			Me.buttonTestImageProxy.Location = New System.Drawing.Point(197, 25)
			Me.buttonTestImageProxy.Name = "buttonTestImageProxy"
			Me.buttonTestImageProxy.Size = New System.Drawing.Size(105, 23)
			Me.buttonTestImageProxy.TabIndex = 0
			Me.buttonTestImageProxy.Text = "Test Image Proxy"
'			Me.buttonTestImageProxy.Click += New System.EventHandler(Me.buttonTestImageProxy_Click);
			' 
			' pictureBox
			' 
			Me.pictureBox.BorderStyle = System.Windows.Forms.BorderStyle.Fixed3D
			Me.pictureBox.Location = New System.Drawing.Point(31, 24)
			Me.pictureBox.Name = "pictureBox"
			Me.pictureBox.Size = New System.Drawing.Size(147, 159)
			Me.pictureBox.TabIndex = 1
			Me.pictureBox.TabStop = False
'			Me.pictureBox.Click += New System.EventHandler(Me.pictureBox_Click);
			' 
			' label1
			' 
			Me.label1.Font = New System.Drawing.Font("Microsoft Sans Serif", 8.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, (CByte(0)))
			Me.label1.Location = New System.Drawing.Point(194, 51)
			Me.label1.Name = "label1"
			Me.label1.Size = New System.Drawing.Size(118, 20)
			Me.label1.TabIndex = 2
			Me.label1.Text = "Click button twice"
'			Me.label1.Click += New System.EventHandler(Me.label1_Click);
			' 
			' FormVirtualProxy
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(323, 217)
			Me.Controls.Add(Me.label1)
			Me.Controls.Add(Me.pictureBox)
			Me.Controls.Add(Me.buttonTestImageProxy)
			Me.Name = "FormVirtualProxy"
			Me.Text = "Virtual Proxy Test"
			CType(Me.pictureBox, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub
		#End Region

		''' <summary>
		''' The main entry point for the application.
		''' </summary>
		<STAThread> _
		Shared Sub Main()
			Application.Run(New FormVirtualProxy())
		End Sub

		Private Sub buttonTestImageProxy_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles buttonTestImageProxy.Click
			Me.pictureBox.Image = New ImageProxy().Image
		End Sub

		Private Sub pictureBox_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles pictureBox.Click

		End Sub

		Private Sub label1_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles label1.Click

		End Sub

		Private Class ImageProxy
			Private Shared _image As Image = Nothing
			Private _width As Integer = 133
			Private _height As Integer = 154
			Private _retrieving As Boolean = False

			Public ReadOnly Property Width() As Integer
				Get
					Return If(_image Is Nothing, _width, _image.Width)
				End Get
			End Property
			Public ReadOnly Property Height() As Integer
				Get
					Return If(_image Is Nothing, _height, _image.Height)
				End Get
			End Property

			Public ReadOnly Property Image() As Image
				Get
					If _image IsNot Nothing Then
						Return _image
					Else
						If (Not _retrieving) Then
							_retrieving = True
							Dim retrievalThread As New Thread(New ThreadStart(AddressOf RetrieveImage))
							retrievalThread.Start()
						End If
						Return PlaceHolderImage()
					End If
				End Get
			End Property

			Public Function PlaceHolderImage() As Image
				Return New Bitmap(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("PlaceHolder.jpg"))
			End Function

			Public Sub RetrieveImage()
				' Book image from amazon
				Dim url As String = "http://images.amazon.com/images/P/0596007124.01._PE34_SCMZZZZZZZ_.jpg"

				Dim request As HttpWebRequest = CType(HttpWebRequest.Create(url), HttpWebRequest)
				Dim response As HttpWebResponse = CType(request.GetResponse(), HttpWebResponse)
				_image = Image.FromStream(response.GetResponseStream())
			End Sub
		End Class
	End Class
End Namespace

File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Proxy.VirtualProxy/FormVirtualProxy.cs
using System;
using System.Drawing;
using System.ComponentModel;
using System.Windows.Forms;
using System.Threading;
using System.Net;

namespace DoFactory.HeadFirst.Proxy.VirtualProxy
{
    /// <summary>
    /// Summary description for FormVirtualProxy.
    /// </summary>
    public class FormVirtualProxy : System.Windows.Forms.Form
    {
        private System.Windows.Forms.Button buttonTestImageProxy;
        private System.Windows.Forms.PictureBox pictureBox;
        private System.Windows.Forms.Label label1;
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;

        public FormVirtualProxy()
        {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose( bool disposing )
        {
            if( disposing )
            {
                if (components != null) 
                {
                    components.Dispose();
                }
            }
            base.Dispose( disposing );
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.buttonTestImageProxy = new System.Windows.Forms.Button();
            this.pictureBox = new System.Windows.Forms.PictureBox();
            this.label1 = new System.Windows.Forms.Label();
            ((System.ComponentModel.ISupportInitialize)(this.pictureBox)).BeginInit();
            this.SuspendLayout();
            // 
            // buttonTestImageProxy
            // 
            this.buttonTestImageProxy.Location = new System.Drawing.Point(197, 25);
            this.buttonTestImageProxy.Name = "buttonTestImageProxy";
            this.buttonTestImageProxy.Size = new System.Drawing.Size(105, 23);
            this.buttonTestImageProxy.TabIndex = 0;
            this.buttonTestImageProxy.Text = "Test Image Proxy";
            this.buttonTestImageProxy.Click += new System.EventHandler(this.buttonTestImageProxy_Click);
            // 
            // pictureBox
            // 
            this.pictureBox.BorderStyle = System.Windows.Forms.BorderStyle.Fixed3D;
            this.pictureBox.Location = new System.Drawing.Point(31, 24);
            this.pictureBox.Name = "pictureBox";
            this.pictureBox.Size = new System.Drawing.Size(147, 159);
            this.pictureBox.TabIndex = 1;
            this.pictureBox.TabStop = false;
            this.pictureBox.Click += new System.EventHandler(this.pictureBox_Click);
            // 
            // label1
            // 
            this.label1.Font = new System.Drawing.Font("Microsoft Sans Serif", 8.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ((byte)(0)));
            this.label1.Location = new System.Drawing.Point(194, 51);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(118, 20);
            this.label1.TabIndex = 2;
            this.label1.Text = "Click button twice";
            this.label1.Click += new System.EventHandler(this.label1_Click);
            // 
            // FormVirtualProxy
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(323, 217);
            this.Controls.Add(this.label1);
            this.Controls.Add(this.pictureBox);
            this.Controls.Add(this.buttonTestImageProxy);
            this.Name = "FormVirtualProxy";
            this.Text = "Virtual Proxy Test";
            ((System.ComponentModel.ISupportInitialize)(this.pictureBox)).EndInit();
            this.ResumeLayout(false);

        }
        #endregion

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main() 
        {
            Application.Run(new FormVirtualProxy());
        }

        private void buttonTestImageProxy_Click(object sender, System.EventArgs e)
        {
            this.pictureBox.Image = new ImageProxy().Image;
        }

        private void pictureBox_Click(object sender, System.EventArgs e)
        {

        }

        private void label1_Click(object sender, System.EventArgs e)
        {

        }

        private class ImageProxy
        {
            private static Image _image = null;
            private int _width = 133;
            private int _height = 154;
            private bool _retrieving = false;

            public int Width
            {
                get{ return _image == null ? _width : _image.Width; }
            }
            public int Height
            {
                get{ return _image == null ? _height : _image.Height; }
            }

            public Image Image
            {
                get
                {
                    if (_image != null)
                        return _image;
                    else
                    {
                        if (!_retrieving)
                        {
                            _retrieving = true;
                            Thread retrievalThread = new Thread(new ThreadStart(RetrieveImage));
                            retrievalThread.Start();
                        }
                        return PlaceHolderImage();
                    }
                }
            }

            public Image PlaceHolderImage()
            {
                return new Bitmap(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("DoFactory.HeadFirst.Proxy.VirtualProxy.PlaceHolder.jpg"));
            }

            public void RetrieveImage()
            {
                // Book image from amazon
                string url = "http://images.amazon.com/images/P/0596007124.01._PE34_SCMZZZZZZZ_.jpg";

                HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(url);
                HttpWebResponse response = (HttpWebResponse)request.GetResponse();
                _image = Image.FromStream(response.GetResponseStream());
            }
        }
    }
}
